src/app/encoder.cpp

```diff
@@ -0,0 +1,170 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "chip-zcl-zpro-codec.h"
+#include <assert.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <support/BufBound.h>
+#include <support/logging/CHIPLogging.h>
+
+using namespace chip;
+extern "C" {
+
+static uint16_t doEncodeApsFrame(BufBound & buf, uint16_t profileID, uint16_t clusterId, uint8_t sourceEndpoint,
+                                 uint8_t destinationEndpoint, EmberApsOption options, uint16_t groupId, uint8_t sequence,
+                                 uint8_t radius)
+{
+
+    uint8_t control_byte = 0;
+    buf.Put(control_byte); // Put in a control byte
+    buf.PutLE16(profileID);
+    buf.PutLE16(clusterId);
+    buf.Put(sourceEndpoint);
+    buf.Put(destinationEndpoint);
+    buf.PutLE(options, sizeof(EmberApsOption));
+    buf.PutLE16(groupId);
+    buf.Put(sequence);
+    buf.Put(radius);
+
+    if (!buf.Fit() && buf.Size() > 0)
+    {
+        ChipLogError(Zcl, "Could not encode buffer due to insufficient space");
+    }
+    else
+    {
+        ChipLogProgress(Zcl, "Successfully encoded APS frame of size %d bytes", buf.Written());
+    }
+
+    return (buf.Fit() || buf.Size() == 0) ? buf.Written() : 0; // 0 size buffer implies measure case.
+}
+
+uint16_t encodeApsFrame(uint8_t * buffer, uint16_t buf_length, EmberApsFrame * apsFrame)
+{
+    if (buffer == nullptr)
+    {
+        buf_length = 0; // This is the measuring case.
+    }
+    BufBound buf = BufBound(buffer, buf_length);
+    return doEncodeApsFrame(buf, apsFrame->profileId, apsFrame->clusterId, apsFrame->sourceEndpoint, apsFrame->destinationEndpoint,
+                            apsFrame->options, apsFrame->groupId, apsFrame->sequence, apsFrame->radius);
+}
+
+uint32_t _encodeOnOffCommand(uint8_t * buffer, uint32_t buf_length, uint8_t command, uint8_t destination_endpoint)
+{
+    uint32_t result = 0;
+    // pick cluster id as 6 for now.
+    // pick source and destination end points as 1 for now.
+    // Profile is 65535 because that matches our simple generated code, but we
+    // should sort out the profile situation.
+
+    if (buffer == nullptr)
+    {
+        buf_length = 0; // This is the measuring case.
+    }
+
+    BufBound buf = BufBound(buffer, buf_length);
+    result       = doEncodeApsFrame(buf, 65535, 6, 1, destination_endpoint, 0, 0, 0, 0);
+    if (result == 0)
+    {
+        ChipLogError(Zcl, "Error encoding aps frame result %" PRIu32 "\n", result);
+        result = 0;
+        return result;
+    }
+
+    // This is a cluster-specific command so low two bits are 0b01.  The command
+    // is standard, so does not need a manufacturer code, and we're sending
+    // client to server, so all the remaining bits are 0.
+    buf.Put(0x1);
+
+    // Transaction sequence number.  Just pick something.
+    buf.Put(0x1);
+
+    buf.Put(&command, sizeof(uint8_t));
```

Comment:
This is going to grab the first byte in memory-layout order from `command`.  That's wrong on BE platforms, right?  I'd rather we didn't add that pitfall, even if we're not running this code on anything BE yet.

Again, we should either change `command` to be a `uint8_t` or assign into a local `uint8_t` here or something....